cdap-app-fabric/src/main/java/co/cask/cdap/internal/provision/ProvisioningService.java

```diff
@@ -384,6 +410,44 @@ Runnable deprovision(ProgramRunId programRunId, DatasetContext datasetContext,
     return createDeprovisionTask(provisioningTaskInfo, provisioner, taskCleanup);
   }
 
+  /**
+   * Checks if a program with the given plugins requirements can run successfully on the provisioner with given
+   * capability  without any incapability failures. Incapability is defined as a provisioner not meeting all the {@link
+   * Requirements} needed by all the plugins in the program.
+   *
+   * @param programName the name of the program to be run
+   * @param provisionerName the name of the provisioner
+   * @param provisionerCapabilities provisioner capabilities
+   * @param pluginRequirements a {@link Map} plugin name to {@link Set} of plugin capabilities
+   *
+   * @throws IncapableException if any of the plugins involved in the program have a requirement which is not met by the
+   * provisioner.
+   */
+  @VisibleForTesting
+  void canRun(String programName, String provisionerName, Set<String> provisionerCapabilities,
+              Map<String, Set<String>> pluginRequirements) throws IncapableException {
+    // create a map of plugin name to unfulfilled requirement if there are any
+    Map<String, Set<String>> unfulfilledRequirements = new HashMap<>();
+    pluginRequirements.forEach((pluginName, pluginRequirement) -> {
+      Set<String> capabilities = provisionerCapabilities.stream().map(String::toLowerCase).collect(Collectors.toSet());
+      Sets.SetView<String> unfulfilledRequirement = Sets.difference(pluginRequirement, capabilities);
+      // if there are any requirement which are unfulfilled by the provisioner then store it
+      if (!unfulfilledRequirement.isEmpty()) {
+        unfulfilledRequirements.put(pluginName, unfulfilledRequirement.immutableCopy());
+      }
+    });
+    if (!unfulfilledRequirements.isEmpty()) {
+      throw new IncapableException(programName, provisionerName, unfulfilledRequirements);
+    }
+  }
+
+  private void canRun(String programName, Provisioner provisioner, Set<Plugin> plugins) throws IncapableException {
+    Map<String, Set<String>> pluginRequirements = new HashMap<>();
+    plugins.forEach(pluginRequirement -> pluginRequirements.put(pluginRequirement.getPluginClass().getName(),
```

Comment:
The plugin name is not unique, as plugins of different types can have the same name. This will end up clobbering each other, possibly allowing a run that should not happen. Make sure this scenario is covered in unit tests.

Reply:
The Map isn't being used for lookups. It's just being used to hold an additional piece of information. Instead of using a Map, create some other class that has the fields you need (type, name, requirements) and pass a collection of those to canRun()